examples/networking/ping_reply.py

```diff
@@ -0,0 +1,137 @@
+        unsigned short seq;
+  } BPF_PACKET_HEADER;
+
+  /*
+  struct __sk_buff is passed by the kernel to this function
+  */
+  int ping_reply (struct __sk_buff * skb) {
+        // Packet is available on the 0th location of the memory.
+        u8 *cursor = 0;
+
+        struct ethernet_t * ethernet = cursor_advance(cursor,
+                sizeof(*ethernet));
+        // If packet type is not IP, return
+        if (ethernet->type != ETH_P_IP)
+                return 0;
+
+        struct ip_t * ip = cursor_advance(cursor, sizeof(*ip));
+        // If next protocol is not ICMP, return
+        if (ip->nextp != IPPROTO_ICMP)
+                return 0;
+
+        struct icmp_t * icmp = cursor_advance(cursor, sizeof(*icmp));
+        // If ICMP packet is not echo, return
+        if (icmp->type != ICMP_ECHO)
+                return 0;
+
+        /*
+        Converting ICMP echo into ICMP reply by changing the type to 0
+        Since we're changing packet contents, we need to update the checksum
+        */
+        unsigned short type = ICMP_ECHOREPLY;
+        bpf_l4_csum_replace(skb,36,icmp->type, type,sizeof(type));
+        bpf_skb_store_bytes(skb, 34, &type, sizeof(type),0);
+
+        /*
+        Swapping Source and Destination in IP header
+        We don't need to update checksum since we're just swapping.
+        However to demonstrate the use of bpf_l3_csum_replace, the checksum
+        is recomputed after each change
+        */
+        unsigned int old_src = bpf_ntohl(ip->src);
+        unsigned int old_dst = bpf_ntohl(ip->dst);
+
+        bpf_l3_csum_replace(skb, IP_CKSUM_OFF, old_src, old_dst,
+                sizeof(old_dst));
+
+        /*
+        Demonstrating the use of bpf_skb_store_bytes(...)
+        There are other easier ways to swap ip->src and ip->dst, one of which
+        follows while swapping mac addresses
+        */
+        bpf_skb_store_bytes(skb, IP_SRC_OFF, &old_dst,
+                sizeof(old_dst), 0);
+
+        bpf_l3_csum_replace(skb, IP_CKSUM_OFF, old_dst, old_src,
+                sizeof(old_src));
+        bpf_skb_store_bytes(skb, IP_DST_OFF, &old_src, sizeof(old_src), 0);
+
+        /* Swapping Mac Addresses
+        Using two temp variables since assigning one memory location
+        to another directly causes a compilation error.
+        */
+        unsigned long long old_src_mac = ethernet->src;
+        unsigned long long old_dst_mac = ethernet->dst;
+
+        ethernet->src = old_dst_mac;
+        ethernet->dst = old_src_mac;
+
+        u64 ret = bpf_clone_redirect(skb, skb->ifindex,0 /*For Egress */);
+        /*
+        This output to the kernel trace_pipe which can also be read by:
+        cat /sys/kernel/debug/tracing/trace_pipe
+        */
+        bpf_trace_printk("Replying to ICMP echo: Result=%u\\n", ret);
+
+        return 1;
+}
+"""
+ipr = IPRoute()
+ipdb = IPDB(nl=ipr)
+ifc = ipdb.interfaces.eth0
+
+b = BPF(text=prog)
+pbr = b.load_func("ping_reply", BPF.SCHED_CLS)
```

Comment:
The convention we are going with is to use SCHED_ACT, and from inside the program you should call bpf_redirect and then return TC_ACT_REDIRECT instead of TC_ACT_DROP. This is more efficient in that it completely avoids cloning the packet.